### PR TITLE
fix: correctly handle HttpError on the client

### LINKED ISSUE #7966: support for throwing `error` in `handle` function
### Describe the problem

When doing a client-side `__data.json` request, errors thrown with the `error` helper function inside `hooks.server.ts` , will not return the correct error object.

It works as expected when `error` gets thrown inside a `+server.*` file.

Here is a demonstration of the issue; https://stackblitz.com/edit/sveltejs-kit-template-default-djcepj?file=src/hooks.server.ts

1. start the application
2. click on the "test hooks.server.ts" link
3. a 401 error gets thrown inside `hooks.servert.ts`
4. the page shows a 500 error
5. reload the page
6. the 401 error gets correctly shown

related: https://github.com/sveltejs/kit/issues/7609
The issue was already described as a sidenote there.

### Describe the proposed solution

`SvelteKit` should support throwing `error` objects also for `__data.json` requests inside `hooks.server.ts`.
This makes the usage of the `error` helper function more consistent.

### Alternatives considered

Leave it as it is

### Importance

would make my life easier

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by gtm-nayan**
Closes #7966
Closes https://github.com/sveltejs/kit/pull/8142

Reading [this comment](https://github.com/sveltejs/kit/pull/8142#issuecomment-1351327046), there's a chance I may have completely misunderstood what the PR was trying to do, but at least this is what I got from the repro and the list of steps on #7966.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

**Comment by dummdidumm** [comment]
The problem I tried to articulate there is that if you throw an error from the handle hook, there's nothing to fall back to. The handle hook is "above" the root layout, which means we need to fall back to the static error page (which you can customize through `error.html`). That happens when hitting the server directly already, but for the client there's no good way to do that currently: If the error is thrown from a `__data.json` request, we can't magically transform that response into the static error page. We can't just show the root layout and its error page either, because the root layout might fire off another `__data.json` request, which results in another error thrown in the handle hook.

**Comment by gtm-nayan** [comment]
Trying to wrap my head around it, but feels like that probably warrants a separate discussion, this change still addresses #7966 right?

**Comment by dummdidumm** [comment]
Sort of - if you have a `+layout.server.js` in your root, that request will still be done when falling back to the root error page with the root layout, which triggers another handle hook invocation. This will probably trigger a full reload and _that_ will show the static error page, but if you're in SPA mode then it won't I think.

This PR makes sense to me in either way because the unexpected error which is thrown while request `__data.json` should've been handled by the server and its `handleError` already, so it shouldn't result in another client `handleError` invocation.

File: packages/kit/src/runtime/client/client.js [hunk 1]
```diff
@@ -1694,7 +1694,7 @@ async function load_data(url, invalid) {
 
 	if (!res.ok) {
 		// error message is a JSON-stringified string which devalue can't handle at the top level
-		throw new Error(data);
+		throw new HttpError(res.status, data);
```
Comment (dummdidumm) [inline packages/kit/src/runtime/client/client.js:1697]:
```suggestion
		// turn it into a HttpError to not call handleError on the client again (was already handled on the server)
		throw new HttpError(res.status, data);
```

**Review by dummdidumm — APPROVED** [review]

**Pull request merged**